tests/test_parameter_interpolation.py

```diff
@@ -83,4 +84,104 @@ def test_nonbonded(self):
+        # are correct.
+
+        np.random.seed(4321)
+        D = 3
+
+        cutoff = 1.0
+        size = 36
+
+        water_coords = self.get_water_coords(D, sort=False)
+        coords = water_coords[:size]
+        padding = 0.2
+        diag = np.amax(coords, axis=0) - np.amin(coords, axis=0) + padding
+        box = np.eye(3)
+        np.fill_diagonal(box, diag)
+
+        N = coords.shape[0]
+
+        lambda_plane_idxs = np.random.randint(low=0, high=2, size=N, dtype=np.int32)
+        lambda_offset_idxs = np.random.randint(low=0, high=2, size=N, dtype=np.int32)
+
+        for precision, rtol in [(np.float64, 1e-8), (np.float32, 1e-4)]:
+
+            # E = 0 # DEBUG!
+            qlj_src, ref_potential, test_potential = prepare_water_system(
+                coords,
+                lambda_plane_idxs,
+                lambda_offset_idxs,
+                p_scale=1.0,
+                cutoff=cutoff
+            )
+
+            qlj_dst, _, _ = prepare_water_system(
+                coords,
+                lambda_plane_idxs,
+                lambda_offset_idxs,
+                p_scale=1.0,
+                cutoff=cutoff
+            )
+
+            def transform_q(lamb):
+                return lamb*lamb
+
+            def transform_s(lamb):
+                return jnp.sin(lamb*np.pi/2)
+
+            def transform_e(lamb):
+                return jnp.cos(lamb*np.pi/2)
+
+            def transform_w(lamb):
+                return (1-lamb*lamb)
+
+            def interpolate_params(lamb, qlj_src, qlj_dst):
+                new_q = (1-transform_q(lamb))*qlj_src[:, 0] + transform_q(lamb)*qlj_dst[:, 0]
+                new_s = (1-transform_s(lamb))*qlj_src[:, 1] + transform_s(lamb)*qlj_dst[:, 1]
+                new_e = (1-transform_e(lamb))*qlj_src[:, 2] + transform_e(lamb)*qlj_dst[:, 2]
+                return jnp.stack([new_q, new_s, new_e], axis=1)
+
+            def u_reference(x, params, box, lamb):
+                d4 = cutoff*(lambda_plane_idxs + lambda_offset_idxs*transform_w(lamb))
+                d4 = jnp.expand_dims(d4, axis=-1)
+                x = jnp.concatenate((x, d4), axis=1)
+
+                qlj_src = params[:len(params)//2]
+                qlj_dst = params[len(params)//2:]
+                qlj = interpolate_params(lamb, qlj_src, qlj_dst)
+                return ref_potential(x, qlj, box, lamb)
+
+
+            for lamb in [0.0, 0.2, 1.0]:
+
+                    qlj = np.concatenate([qlj_src, qlj_dst])
+
+                    print("lambda", lamb, "cutoff", cutoff, "precision", precision, "xshape", coords.shape)
+
+                    args = copy.deepcopy(test_potential.args)
+                    args.append("lambda*lambda") # transform q
+                    args.append("sin(lambda*PI/2)") # transform sigma
+                    args.append("cos(lambda*PI/2)") # transform epsilon
+                    args.append("1-lambda*lambda") # transform w
```

Comment:
> Are there any other global constants that can be referenced in these expressions aside from PI?

Right now no, but you can always just hard code in the constants. It woudln't hard to add in other constants though.

> Is it possible to define and reuse variables here? (Something like 0.1 * x + 0.2 * x*x + 0.3 * x*x*x ; x=(1 - lambda*lambda);)

Yes absolutely, but right now the C++ code is written as `return CUSTOM_EXPRESSION`; to support proper branching (eg. if statements beyond simple ternary operators). Currently it's similar to lambda expressions, but  we can definitely relax this without difficulty.

> Is it possible to reference per-atom attributes in these expressions?

Currently no, but if these are forcefield independent attributes, we may be able to support them without too much difficulty. We probably won't be able to do derivatives for the per-atom attributes though, since the forward-mode AD/CSD is only efficient for `R^1->R^N`.

> Is there a practical limit on the length of these expressions?

Anything that doesn't break the compiler :)